### PR TITLE
Print errors properly

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: pil-analyzer/src/type_inference.rs [hunk 1]
```diff
@@ -592,24 +595,29 @@ impl<'a> TypeChecker<'a> {
...
             }
             Expression::FreeInput(_, _) => todo!(),
-            Expression::MatchExpression(_, MatchExpression { scrutinee, arms }) => {
+            Expression::MatchExpression(source_ref, MatchExpression { scrutinee, arms }) => {
                 let scrutinee_type = self.infer_type_of_expression(scrutinee)?;
                 let result = self.new_type_var();
                 for MatchArm { pattern, value } in arms {
                     let local_var_count = self.local_var_types.len();
-                    self.expect_type_of_pattern(&scrutinee_type, pattern)?;
+                    // TODO we need a better source reference.
```
Comment (chriseth) [inline pil-analyzer/src/type_inference.rs:618]: There is already a PR for it.

File: pil-analyzer/src/type_inference.rs [hunk 2]
```diff
@@ -713,11 +722,12 @@ impl<'a> TypeChecker<'a> {
         self.unifier
             .unify_types(inferred_type.clone(), expected_type.clone())
             .map_err(|err| {
-                format!(
-                    "Error checking sub-expression {expr}:\nExpected type: {}\nInferred type: {}\n{err}",
+                // TODO we should also list the trait constraints.
```
Comment (chriseth) [inline pil-analyzer/src/type_inference.rs:725]: Maybe we should still do this, but could also be done in a different PR.

File: pil-analyzer/tests/types.rs [hunk 3]
```diff
@@ -156,6 +156,8 @@ fn generic_fixes_concrete() {
     // but in order to do that, we have to unify the derived type of
     // `y` with the declared and this would maybe create problems in that
     // we would not derive the most generic type for generic functions.
+
+    // TODO revisit this test, check if we are missing an error.
```
Comment (chriseth) [inline pil-analyzer/tests/types.rs:160]: Still TODO.

File: pil-analyzer/src/type_inference.rs [hunk 4]
```diff
@@ -58,6 +59,8 @@ struct TypeChecker<'a> {
     /// Declared types for all symbols. Contains the unmodified type scheme for symbols
     /// with generic types and newly created type variables for symbols without declared type.
     declared_types: HashMap<String, TypeScheme>,
+    /// Source references for all declared types.
+    declared_type_source_refs: HashMap<String, SourceRef>,
```
Comment (leonardoalt) [inline pil-analyzer/src/type_inference.rs:63]: is this supposed to have the same keys as `declared_types`? If yes, these could be a tuple here

Comment (chriseth) [reply pil-analyzer/src/type_inference.rs:63]: Done.

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**